src/_macosx.m

```diff
@@ -178,6 +178,101 @@ static int wait_for_stdin(void)
     return 1;
 }
 
+static PyObject* set_icon(PyObject* ignored, PyObject* path)
+{
+    if (!path)
+    {
+        PyErr_SetString(PyExc_RuntimeError, "No path passed");
+        return NULL;
+    }
+
+    const char* path_string;
+    PyObject* unicode_path = PyObject_Str(path);
+    if (!unicode_path)
+    {
+        return NULL;
+    }
+
+    // API indicates Python owns path_string; we don't need to free it
+    path_string = PyUnicode_AsUTF8(unicode_path);
+    Py_DECREF(unicode_path);
+    if (!path_string)
+    {
+        PyErr_SetString(PyExc_TypeError, "bad argument type, must be "
```

Comment:
you probably don't need to set the exception, PyUnicode_AsUTF8 will already have done it for you

Reply:
I set this one because the exception they use is opaque. IIRC, it’s something like, “bad argument type in compiled code”. I was iffy on setting my own? I could go either way, let me know.

Reply:
See latest push.